_posts/2024-04-23-kubernetes-horizontal-pod-autoscaler-guide.md

```diff
@@ -0,0 +1,170 @@
+
+<div class="post-center-image">
+    {% picture pimage /assets/images/fullsize/posts/2024-04-23-kubernetes-horizontal-pod-autoscaler-guide/cpu_stairs.png --alt HPA CPU expected replicas showing stairs pattern %}
+</div>
+
+{:refdef: class="image-caption"}
+*Figure 4. HPA CPU current replicas vs expected replicas*
+{: refdef}
+
+There you go! We can clearly see the staircase pattern, similar to $f(x) = x + 1$. This occurs because the `currentMetricValue / desiredMetricValue` ratio is greater than 1, prompting the autoscaler to continuously increase the number of replicas until it reaches the maximum allowed. In this instance, `maxReplicas` was set to 3. Thus, we have identified the root of the problem!
+
+## Why is HPA not scaling down?
+
+Although CPU usage spiked at startup, it quickly returned to low levels. So why isn't the HPA scaling down the number of replicas? It appears that the CPU requirement is well below the target value of `450m`, as illustrated in *Figure 1*. According to the official HPA documentation:
+
+> "If multiple metrics are specified in a HorizontalPodAutoscaler, this calculation is done for each metric, and then the largest of the desired replica counts is chosen."
+
+This indicates that the issue now lies with memory usage. Grafana shows us that memory usage has remained constant after the scaling. According to *Figure 2*, the expected number of replicas should be just 1. However, since the HPA previously increased our replicas to 3, when we view the same plot with the number of replicas on the x-axis, it reveals the following:
+
+<div class="post-center-image">
+    {% picture pimage /assets/images/fullsize/posts/2024-04-23-kubernetes-horizontal-pod-autoscaler-guide/memory_stairs.png --alt HPA memory current replicas vs expected replicas %}
+</div>
+
+{:refdef: class="image-caption"}
+*Figure 5. HPA memory current replicas vs expected replicas*
+{: refdef}
+
+With the current memory usage, the HPA behaves like the function $f(x) = x$, preventing the number of replicas from scaling down. This is why we continuously see three pods running, even though the pods are not receiving much traffic.
+
+## What can we do?
+
+We have a couple of options to address this problem. For instance, we could change the memory and CPU targets in the HPA settings. Increasing the CPU requirements could help us to avoid the initial spike in replicas and prevent the staircase pattern. On the other hand if we increase the memory requirements we could avoid the bottleneck and scale down the number of replicas. However, this isn't a permanent solution because if our application's memory use fluctuates, we could encounter the same issue again. Instead, we should tackle the root cause and for this we need to differentiate between the fixed and variable usage of our metrics.
```

Comment:
```suggestion
We have a couple of options to address this problem. For instance, we could change the memory and CPU targets in the HPA settings. Increasing the CPU requirements could help us avoid the initial spike in replicas and prevent the staircase pattern. On the other hand, if we increase the memory requirements, we could avoid the bottleneck and scale down the number of replicas. However, this isn't a permanent solution because if our application's memory use fluctuates, we could encounter the same issue again. Instead, we should tackle the root cause and for this we need to differentiate between the fixed and variable usage of our metrics.
```